src/mem/arena.rs

```diff
@@ -337,14 +361,135 @@ unsafe impl<'arena> Arena for BumpArena<'arena> {
     }
 }
 
+/// A type that can serve as a buffer an arena can allocate from.
```

Comment:
There's a couple of issues:
- AsRef does not guarantee that the returned slice is stable assuming the self value isn't moved.
- Holding onto a pointer into the storage doesn't work, because if the storage is inline, moving the arena will invalidate it. There's no way to detect this.

Reply:
Wouldn't the borrow checker prevent this from occurring? How would you move a `BumpArena` while references with lifetime `'arena` are used?

Reply:
I made a [concrete example](https://gist.github.com/maurer/14be4a82ce87293b96883a675b0bdc67) of what I meant here.

The logic as to why this works is that the `BumpArenaRef` has a lifetime restricted to that of the arena, and in turn anything allocated from that ref is restricted, so the arena should not be movable while any of these allocations are out.

Reply:
Here's the problem, as I see it:

- I can't save the pointer to the storage in the struct at creation time, because if the struct moves, that pointer dangles.
- I can't call AsMut as necessary, because that requires an `&mut T`, which I can't create after handing out memory without violating the aliasing rules. This is why your example doesn't work, Matt.

Since neither of these solutions work, what I'm currently doing is deconstructing the storage into something I can safely get a raw pointer and a length out of, at which point I can offset things around and produce a truly unique slice reference.